packages/cli/src/executionLifecycleHooks/shared/sharedHookFunctions.ts

```diff
@@ -0,0 +1,99 @@
+import type { ExecutionStatus, IRun, IWorkflowBase } from 'n8n-workflow';
+import type { IExecutionDb } from '@/Interfaces';
+import pick from 'lodash/pick';
+import { isWorkflowIdValid } from '@/utils';
+import { LoggerProxy } from 'n8n-workflow';
+import Container from 'typedi';
+import { ExecutionRepository } from '../../databases/repositories';
+import { ExecutionMetadataService } from '../../services/executionMetadata.service';
+
+export function determineFinalExecutionStatus(runData: IRun): ExecutionStatus {
+	const workflowHasCrashed = runData.status === 'crashed';
+	const workflowWasCanceled = runData.status === 'canceled';
+	const workflowDidSucceed =
+		!runData.data.resultData.error && !workflowHasCrashed && !workflowWasCanceled;
+	let workflowStatusFinal: ExecutionStatus = workflowDidSucceed ? 'success' : 'failed';
+	if (workflowHasCrashed) workflowStatusFinal = 'crashed';
+	if (workflowWasCanceled) workflowStatusFinal = 'canceled';
+	if (runData.waitTill) workflowStatusFinal = 'waiting';
+	return workflowStatusFinal;
+}
```

Comment:
I can't wait to fix the hooks and get rid of this function.
```suggestion
// TODO: get rid of this function - the status needs to be set somewhere else according to execution results and shouldn't be inferred here.
export function determineFinalExecutionStatus(runData: IRun): ExecutionStatus {
	const workflowHasCrashed = runData.status === 'crashed';
	const workflowWasCanceled = runData.status === 'canceled';
	const workflowDidSucceed =
		!runData.data.resultData.error && !workflowHasCrashed && !workflowWasCanceled;
	let workflowStatusFinal: ExecutionStatus = workflowDidSucceed ? 'success' : 'failed';
	if (workflowHasCrashed) workflowStatusFinal = 'crashed';
	if (workflowWasCanceled) workflowStatusFinal = 'canceled';
	if (runData.waitTill) workflowStatusFinal = 'waiting';
	return workflowStatusFinal;
}
```